File: client.Angular/src/app/smart-geo-iot/Device.ts
import { Project, ProjectEnum } from './project';
import { Package } from './package';
import { Bits } from './Bits';

export class Device {
    id: string;
    name: string;
    sequenceNumber: number;
    lastCom: number;
    state: number;
    comState: number;
    pac: string;
    locationLat: string;
    locationLng: string;
    deviceTypeId: string;
    groupId: string;
    lqi: number;
    activationTime: number;
    tokenState: number;
    tokenDetailMessage: string;
    tokenEnd: number;
    contractId: string;
    creationTime: number;
    modemCertificateId: string;
    prototype: boolean;
    automaticRenewal: boolean;
    automaticRenewalStatus: number;
    createdBy: string;
    lastEditionTime: number;
    lastEditedBy: string;
    activable: boolean;
    bits: Bits;

    get getDeviceNameId(): string {
        return this.id + ' (' + this.name + ')';
    }
}

export class DeviceRegistration {
    deviceCustomUId: string;
    name: string;
    nickName: string;
    deviceId: string;
    device: Device;
    packageUId: string;
    package: Package;
    projectUId: string;
    project: Project;
    model: string;
    serialNumber: string;
    notes: string;

    get getUrlRouterLinkProject(): string {
        if (this.project.code === ProjectEnum.B972) {
            return ('clampon/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B978) {
            return ('b978/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B980) {
            return ('b980/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B982_S) {
            return ('aguamon83/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B987) {
            return ('b987/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B975) {
            return ('b975/' + this.deviceId);
        }

        if (this.package.type === '10') {
            return ('aguamon/' + this.deviceId);
        }
        if (this.package.type === '12') {
            return ('djrf/' + this.deviceId);
        }
        if (this.package.type === '81') {
            return ('aguamon81/' + this.deviceId);
        }
        
        if (this.package.type === '21') {
            return ('aguamon21/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B972_P) {
            return ('trm23/' + this.deviceId);
        }
        
        return ('' + this.deviceId);
    }

    get getGraphicUrlRouterLinkProject(): string {
        if (this.project.code === ProjectEnum.B972) {
            // return ('trm23/' + this.deviceId);
            return '/';
        }
        // if (this.project.code === ProjectEnum.B978) {
        //     return ('b978/' + this.deviceId);
        // }
        if (this.project.code === ProjectEnum.B978) {
            return ('trm/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B980) {
            return ('trm/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B987) {
            return ('b987/' + this.deviceId);
        }
        if (this.project.code === ProjectEnum.B975) {
            return ('b975/' + this.deviceId);
        }
        
        if (this.package.type === '10') {
            return ('aguamon/' + this.deviceId);
        }
        if (this.package.type === '12') {
            return ('djrf/' + this.deviceId);
        }
        if (this.package.type === '81') {
            return ('pqa/' + this.deviceId);
        }
        if (this.package.type === '83') {
            return ('tsp/' + this.deviceId);
        }
        if (this.package.type === '21') {
            return ('trm/' + this.deviceId);
        }
        if (this.package.type === '23') {
            return ('trm23/' + this.deviceId);
        }
        return (this.deviceId);
    }
}

export class DeviceLocation {
    deviceLocationUId: string;
    deviceId: string;
    time: number;
    data: string;
    radius: string;
    latitude: number;
    longitude: number;
}

export class B975DevicesDashboardViewModels {
    deviceId: string;
    name: string;
    nickName: string;
    date: Date;
    statusDJ: string;
    contadorCarencias: number;
    contadorBloqueios: number;
    locationCity: string;
    hasServiceDeskOpened: boolean;
    hasHistoryServiceDesk: boolean;
}

File: server/SmartGeoIot/Models/Device.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using SmartGeoIot.Extensions;
using SmartGeoIot.ViewModels;

namespace SmartGeoIot.Models
{
    public class Device
    {
        [Key]
        public string Id { get; set; }
        public string Name { get; set; }
        public int SequenceNumber { get; set; }
        public long LastCom { get; set; }
        public int State { get; set; }
        public int ComState { get; set; }
        public string Pac { get; set; }
        public string LocationLat { get; set; }
        public string LocationLng { get; set; }
        public string DeviceTypeId { get; set; }
        public string GroupId { get; set; }
        public int Lqi { get; set; }
        public long ActivationTime { get; set; }
        public int TokenState { get; set; }
        public string TokenDetailMessage { get; set; }
        public long TokenEnd { get; set; }
        public string ContractId { get; set; }
        public long CreationTime { get; set; }
        public string ModemCertificateId { get; set; }
        public bool Prototype { get; set; }
        public bool AutomaticRenewal { get; set; }
        public int AutomaticRenewalStatus { get; set; }
        public string CreatedBy { get; set; }
        public long LastEditionTime { get; set; }
        public string LastEditedBy { get; set; }
        public bool Activable { get; set; }

        [NotMapped]
        public Bits Bits { get; set; }
    }

    public class DeviceRegistration
    {

        [Key, Column(TypeName = "char(36)"), MaxLength(36)]
        public string DeviceCustomUId { get; set; }

        [MaxLength(250)]
        public string Name { get; set; }

        [MaxLength(100)]
        public string NickName { get; set; }

        // [MaxLength(50)]
        public string DeviceId { get; set; }

        [ForeignKey("DeviceId")]
        public Device Device { get; set; }


        public string PackageUId { get; set; }

        [ForeignKey("PackageUId")]
        public Package Package { get; set; }


        public string ProjectUId { get; set; }

        [ForeignKey("ProjectUId")]
        public Project Project { get; set; }

        [MaxLength(50)]
        public string DataDownloadLink { get; set; }

        [MaxLength(80)]
        public string SerialNumber { get; set; }
        
        [MaxLength(50)]
        public string Model { get; set; }
        public string Notes { get; set; }
        public DateTime? NotesCreateDate { get; set; }
        public string Ed1 { get; set; }
        public string Ed2 { get; set; }
        public string Ed3 { get; set; }
        public string Ed4 { get; set; }
        public string Sd1 { get; set; }
        public string Sd2 { get; set; }
        public string Ea10 { get; set; }
        public string Sa3 { get; set; }

        [NotMapped]
        public string Envio
        {
            get
            {
                if (this.DataDownloadLink == null)
                    return string.Empty;

                return Utils.HexaToDecimal(this.DataDownloadLink.Substring(10, 2)).ToString();
            }
        }

        [NotMapped]
        public string PeriodoTransmissao
        {
            get
            {
                if (this.DataDownloadLink == null)
                    return string.Empty;

                return Utils.HexaToDecimal(this.DataDownloadLink.Substring(4, 4)).ToString();
            }
        }

        [NotMapped]
        public bool BaseTempoUpLink
        {
            get
            {
                if (this.DataDownloadLink == null)
                    return false;

                var _chars = Utils.StringToHexadecimalTwoChars(this.DataDownloadLink);
                var _bytes = Utils.ByteToBinary(_chars[1]);
                return Convert.ToBoolean(int.Parse(_bytes.Substring(7, 1)));
            }
        }

        [NotMapped]
        public string TensaoMinima
        {
            get
            {
                if (this.DataDownloadLink == null)
                    return string.Empty;

                return Utils.HexaToDecimal(this.DataDownloadLink.Substring(8, 2)).ToString();
            }
        }
    }

    public class DeviceType
    {
        public string id { get; set; }
        public string downlinkDataString { get; set; }
    }

    public class DeviceLocation
    {
        [Key, Column(TypeName = "char(36)"), MaxLength(36)]
        public string DeviceLocationUId { get; set; }
        public string DeviceId { get; set; }
        public long Time { get; set; }

        [MaxLength(50)]
        public string Data { get; set; }

        [MaxLength(50)]
        public string Radius { get; set; }

        [MaxLength(100)]
        public double Latitude { get; set; }

        [MaxLength(100)]
        public double Longitude { get; set; }
    }
    
}
